src/snapshot/SnapshotServer.cpp

```diff
@@ -84,18 +83,25 @@ std::unique_ptr<google::protobuf::Message> SnapshotServer::recvPushSnapshot(
       faabric::snapshot::getSnapshotRegistry();
 
     // Set up the snapshot
-    faabric::util::SnapshotData data;
-    data.size = r->contents()->size();
+    faabric::util::SnapshotData snap;
+    snap.size = r->contents()->size();
+
+    if (r->maxSize() > 0) {
+        // Allocate virtual memory big enough for the max size if provided
+        snap.data = faabric::util::allocateVirtualMemory(r->maxSize());
+
+        // Claim just the snapshot region
+        faabric::util::claimVirtualMemory(snap.data, snap.size);
+    } else {
+        // Normal provisioning of just snapshot region
+        snap.data = faabric::util::allocatePrivateMemory(snap.size);
+    }
 
-    // TODO - avoid this copy by changing server superclass to allow subclasses
-    // to provide a buffer to receive data.
-    // TODO - work out snapshot ownership here, how do we know when to delete
-    // this data?
-    data.data = (uint8_t*)mmap(
-      nullptr, data.size, PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-    std::memcpy(data.data, r->contents()->Data(), data.size);
+    // TODO - somehow avoid this copy?
+    std::memcpy(snap.data, r->contents()->Data(), snap.size);
 
-    reg.takeSnapshot(r->key()->str(), data, true);
+    // TODO - avoid a further copy in here when setting up fd
+    reg.takeSnapshot(r->key()->str(), snap, true);
```

Comment:
At least one copy could be avoided here by adding an API to create a snapshot (including its memfd) of a given size without data, then making the caller responsible for filling in the data. The call would have to create the memfd, ftruncate it and mmap it, saving the mmap location in the `data` field

Reply:
Yes that might work, I'll see what I can do.